samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,584 @@
+
+			broker4->sin_addr.s_addr =
+				((struct sockaddr_in *)addr->ai_addr)
+				->sin_addr.s_addr;
+			broker4->sin_family = AF_INET;
+			broker4->sin_port = htons(8883);
+			printk("IPv4 Address found 0x%08x\n",
+			       broker4->sin_addr.s_addr);
+			break;
+		} else {
+			printk("ai_addrlen = %u should be %u or %u\n",
+			       (unsigned int)addr->ai_addrlen,
+			       (unsigned int)sizeof(struct sockaddr_in),
+			       (unsigned int)sizeof(struct sockaddr_in6));
+		}
+
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+static int provision(void)
+{
+	int err;
+
+	err = tls_credential_add(NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_CA_CERTIFICATE,
+					AmazonRootCA3_pem,
+					AmazonRootCA3_pem_len);
+	if (err < 0) {
+		printk("Failed to register ca certificate: %d",
+		       err);
+		return err;
+	}
+	err = tls_credential_add(NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_PRIVATE_KEY,
+				 private_pem_key,
+				 private_pem_key_len);
+	if (err < 0) {
+		printk("Failed to register private key: %d",
+		       err);
+		return err;
+	}
+	err = tls_credential_add(NRF_CLOUD_SEC_TAG,
+				 TLS_CREDENTIAL_SERVER_CERTIFICATE,
+				 public_pem,
+				 public_pem_len);
+	if (err < 0) {
+		printk("Failed to register public certificate: %d",
+		       err);
+		return err;
+	}
+
+	return 0;
+}
+
+static int provision_modem_key(void)
+{
+	static sec_tag_t sec_tag_list[] = {NRF_CLOUD_SEC_TAG};
+
+#if defined(CONFIG_BSD_LIBRARY)
+	{
+		int err;
+
+		/* Delete certificates */
+		nrf_sec_tag_t sec_tag = NRF_CLOUD_SEC_TAG;
+
+		for (nrf_key_mgnt_cred_type_t type = 0; type < 3; type++) {
+			err = nrf_inbuilt_key_delete(sec_tag, type);
+			printk("nrf_inbuilt_key_delete(%lu, %d) => result=%d",
+				sec_tag, type, err);
+		}
+
+		/* Provision CA Certificate. */
+		err = nrf_inbuilt_key_write(NRF_CLOUD_SEC_TAG,
+					NRF_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+					AmazonRootCA3_pem,
+					AmazonRootCA3_pem_len);
+		if (err) {
+			printk("NRF_CLOUD_CA_CERTIFICATE err: %d", err);
+			return err;
+		}
+
+		/* Provision Private Certificate. */
+		err = nrf_inbuilt_key_write(
+			NRF_CLOUD_SEC_TAG,
+			NRF_KEY_MGMT_CRED_TYPE_PRIVATE_CERT,
+			private_pem_key,
+			private_pem_key_len);
+		if (err) {
+			printk("NRF_CLOUD_CLIENT_PRIVATE_KEY err: %d", err);
+			return err;
+		}
+
+		/* Provision Public Certificate. */
+		err = nrf_inbuilt_key_write(
+			NRF_CLOUD_SEC_TAG,
+			NRF_KEY_MGMT_CRED_TYPE_PUBLIC_CERT,
+				 public_pem,
+				 public_pem_len);
+		if (err) {
+			printk("NRF_CLOUD_CLIENT_PUBLIC_CERTIFICATE err: %d",
+				err);
+			return err;
+		}
+	}
+#else
+  #error "Missing CONFIG_BSD_LIB"
+#endif
+}
+
+/* Function to get the client id */
+static int client_id_get(char *id)
+{
+#if !defined(NRF_CLOUD_CLIENT_ID)
+#if defined(CONFIG_BSD_LIBRARY)
+	int at_socket_fd;
+	int bytes_written;
+	int bytes_read;
+	char imei_buf[IMEI_LEN + 1];
+	int ret;
+
+	at_socket_fd = nrf_socket(NRF_AF_LTE, 0, NRF_PROTO_AT);
```

Comment:
There's `at_cmd` module available since recently so that you do not need to handle sockets yourself, but I guess there might be no time to get familiar with it now.

But please, do not use `nrf_socket` API directly from the app, use regular socket API instead - see https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/b7fb4ee10faa2e9b6a8a216fe6ebc7052ae41f3f/drivers/at_cmd/at_cmd.c#L52. Other calls are pretty straightforward, just remove `nrf_` prefix.

Reply:
For some reason I can't use the regular socket API as it gets disabled from `bsd_lib` is this correct or am I doing something wrong.

```
#if !defined(CONFIG_NET_SOCKETS_OFFLOAD)
static inline int socket(int family, int type, int proto) 
```
while
`BSD_LIB selects NET_SOCKETS_OFFLOAD`

Tried to change to the `at_cmd` library but I get a fault when I try to do the same I'll see if I can get it working. It's in the latest commit.

Reply:
In case of `NET_SOCKETS_OFFLOAD` enabled, the APIs are defined in a different header - we use it already in different libraries/applications and it's fine.

But if you get trouble getting it running for whatever reason, let's drop it for now, as we have **very** limited time now :)

Reply:
Well I think I got it working my response buffer was just too small :+1:

Reply:
Got it working with `at_cmd_write` so I'll close this.